chain/chain/src/flat_storage_creator.rs

```diff
@@ -0,0 +1,146 @@
+    /// During this step, we save current chain head, start saving all deltas for blocks after chain head and wait until
+    /// final chain head moves after saved chain head.
+    SavingDeltas,
+    /// We can start fetching state to fill flat storage for some final chain head, because all deltas after it are
+    /// saved to disk. It is done in `NUM_PARTS` / `PART_STEP` steps, during each step we spawn background threads to
+    /// fill some part of state.
+    /// Status contains block hash for which we fetch the shard state and step of fetching state. Progress of each step
+    /// is saved to disk, so if creation is interrupted during some step, it won't repeat previous steps and will start
+    /// from this step again.
+    #[allow(unused)]
+    FetchingState((CryptoHash, u64)),
+    /// Flat storage is initialized but its head is too far away from chain final head. We need to apply deltas until
+    /// the head reaches final head.
+    #[allow(unused)]
+    CatchingUp,
+    /// Flat storage head is the same as chain final head. We can create `FlatStorageState`.
+    #[allow(unused)]
+    Finished,
+}
+
+/// Creates flat storage and tracks creation status for the given shard.
+pub struct FlatStorageShardCreator {
+    pub status: CreationStatus,
+    pub shard_id: ShardId,
+    /// Tracks number of traversed state parts during a single step.
+    #[allow(unused)]
+    pub traversed_state_parts: Option<u64>,
+    /// Used by threads which traverse state parts to tell that traversal is finished.
+    #[allow(unused)]
+    pub traversed_parts_sender: Sender<u64>,
+    /// Used by main thread to update the number of traversed state parts.
+    #[allow(unused)]
+    pub traversed_parts_receiver: Receiver<u64>,
+}
+
+impl FlatStorageShardCreator {
+    pub fn new(shard_id: ShardId, _chain_store: &ChainStore) -> Self {
+        let (traversed_parts_sender, traversed_parts_receiver) = unbounded();
+        // TODO: read flat storage data and set correct status. ChainStore will be used to get flat storage heads and
+        // block heights.
+        let status = CreationStatus::SavingDeltas;
+        Self {
+            status,
+            shard_id,
+            traversed_state_parts: None,
+            traversed_parts_sender,
+            traversed_parts_receiver,
+        }
+    }
+}
+
+/// Creates flat storages for all shards.
+pub struct FlatStorageCreator {
+    /// Height on top of which this struct was created.
+    pub start_height: BlockHeight,
+    pub shard_creators: Vec<Arc<Mutex<FlatStorageShardCreator>>>,
+    pub runtime_adapter: Arc<dyn RuntimeAdapter>,
+    /// Used to spawn threads for traversing state parts.
+    pub pool: rayon::ThreadPool,
+}
+
+impl FlatStorageCreator {
+    pub fn new(runtime_adapter: Arc<dyn RuntimeAdapter>, chain_store: &ChainStore) -> Option<Self> {
+        let chain_head = chain_store.head().unwrap();
+        let num_shards = runtime_adapter.num_shards(&chain_head.epoch_id).unwrap();
```

Comment:
If the migration is already done on a node, when the node restarts, how does it avoid initiating the migration process again?

Reply:
I plan to resolve this in `FlatStorageShardCreator::new` by reading data from disk.
The idea is that:
* if FS head is not stored at all - we set `SavingDeltas` status
* if FS head is stored together with last saved state part id - we set `FetchingState` status
* if FS head is stored alone but is behind final head - we set `CatchingUp` status
* if FS head = final head, we create flat storage state and don't run this process.
This is also described in comments for `CreationStatus`.

Reply:
I see, sorry that I missed the comments. Overall I think this PR looks great!